Kernel/Random.h

```diff
@@ -26,10 +27,98 @@
 
 #pragma once
 
+#include <AK/Assertions.h>
+#include <AK/ByteBuffer.h>
 #include <AK/Types.h>
+#include <Kernel/StdLib.h>
+#include <LibCrypto/Cipher/Cipher.h>
 
 namespace Kernel {
 
+template<typename CipherT, typename HashT, int KeySize>
+class FortunaPRNG {
+    constexpr static size_t pool_count = 32;
+    constexpr static size_t reseed_threshold = 16;
+
+    using CipherType = CipherT;
+    using BlockType = CipherT::BlockType;
+    using HashType = HashT;
+    using DigestType = HashT::DigestType;
+
+public:
+    void get_random_bytes(u8* buffer, size_t n)
+    {
+        if (m_p0_len >= reseed_threshold) {
+            this->reseed();
+        }
+
+        ASSERT(m_counter != 0);
+
+        //FIXME: more than 2^20 bytes cannot be generated without refreshing the key
+        ASSERT(n < (1 << 20));
+
+        CipherType cipher(m_key, m_key.size());
+
+        size_t block_size = CipherType::BlockSizeInBits / 8;
+
+        for (size_t i = 0; i < n; i += block_size) {
+            this->generate_block(cipher, &buffer[i], min(block_size, n - i));
+        }
+
+        // generate a new key
+
+        for (size_t i = 0; i < KeySize; i += block_size) {
+            this->generate_block(cipher, &(m_key[i]), min(block_size, KeySize - i));
+        }
+    }
+
+    template<typename T>
+    void add_random_event(const T& event_data, size_t pool)
+    {
+        pool %= pool_count;
+        if (pool == 0) {
+            m_p0_len++;
+        }
+        m_pools[pool].update(reinterpret_cast<u8*>(&event_data), sizeof(T));
+    }
+
+private:
+    void generate_block(CipherType cipher, u8* buffer, size_t size)
+    {
+        BlockType input((u8*)m_counter, sizeof(m_counter));
+        BlockType output;
+        cipher.encrypt_block(input, output);
+        m_counter++;
```

Comment:
:thinking: 

This looks like CTR mode, if it has all the implications of CTR mode, you could just move these stuff into `Mode/CTR`?
Feel free to stub out the decrypt operation if you do.

Reply:
Done! and as a side effect, `get_random_bytes` is way more elegant. However, in order to use the CTR mode, we now need to create an `AK::ByteBuffer` that is the same length as requested by the user.

I can think of a few solutions to this:

1) Add an API specific to CTR mode that does not encrypt anything.
2) Make the input argument to `Mode::encrypt` optional.
3) Treat empty input to `CTR::encrypt` specially.
4) Just ignore it until it becomes a problem.

Reply:
hmmm...I see how that might be a little verbose to use, but honestly...isn't all of crypto too damn verbose?
mode-specific APIs are fine, if you want to do that, go ahead.
but for the time being, I don't think it's strictly necessary.

Reply:
I am more concerned by the extra allocation (which could be quite large), rather than the verbosity. I'll probably add a specific API.

Reply:
I have added the `key_stream` method to `CTR` which takes no input buffer, and amended my commit.